modules/openfoam/src/OpenFOAM.f90

```diff
@@ -1004,42 +1005,137 @@ END SUBROUTINE CalcForceActuatorPositionsTower
+      !end do
+   end if
+
+
+   ! If non-uniform spacing is called for, compute the spacing.  
+   ! We know that if the spacing is proportional to chord, then at each
+   ! element, ds = a*c, where a is some constant factor across the entire
+   ! blade.  Therefore sum(ds) = sum(a*c) = L, so a is L/sum(c).  Because
+   ! we don't know c at every point exactly beforehand, we have to
+   ! iterate using c from our previous best guess.  We know things have
+   ! converged when a = c/ds is constant along the blade, so that is our
+   ! convergence check.  We take the difference between a = c/ds between
+   ! all neighboring points to see how different they are, and take the
+   ! rms of that error as the convergence measure (eSum).
+   if (p_OpFM%NodeClusterType .eq. 1) then
+
+       ! For chord-based clustering (increase resolution in regions of decreased chord), an iterative solution to the grid spacing is used.
+       ! The initial guess to the spacing is uniform spacing, so start with that.
+       pNonUniform(:) = pUniform(:)
+
+       ! Get the chord at the initial force points.
+       call OpFM_InterpolateForceNodesChord(initOut_AD, p_OpFM, u_OpFM, ErrStat2, ErrMsg2)
+       cNonUniform(1:p_OpFM%nNodesForceBlade) = u_OpFM%forceNodesChord(2:p_OpFM%nNodesForceBlade+1)
+
+       ! Iterate on a chord-based non-uniform spacing.
+       counter = 0
+       e(:) = 1.0e+6
+       eTol = 1.0e-6
+       eSum = sqrt(sum(e(:) * e(:)))
+       do while ((eSum .gt. eTol) .and. (counter < 50))
+
+          !set the non-uniform spacing to ds = (sum(ds^) / sum(c^)) * c^, where
+          !the ^ denotes from the last iteration.  To begin the iteration, we
+          !use ds = uniform.
+          sNonUniform(:) = (p_OpFM%BladeLength)*cNonUniform(:)/(sum(cNonUniform(2:p_OpFM%nNodesForceBlade-1)) + 0.5*(cNonUniform(1)+cNonUniform(p_OpFM%nNodesForceBlade)))
+
+          ! set the new blade points based on the new ds.
+          do i = 2, p_OpFM%nNodesForceBlade
+             pNonUniform(i) = pNonUniform(i-1) + 0.5*(sNonUniform(i-1) + sNonUniform(i))
+          end do
+          pNonUniform(p_OpFM%nNodesForceBlade) = p_OpFM%BladeLength
+          p_OpFM%forceBldRnodes(:) = pNonUniform(:)
+
+          ! interpolate chord to the new points to get the updated chord values
+          call OpFM_InterpolateForceNodesChord(initOut_AD, p_OpFM, u_OpFM,ErrStat2, ErrMsg2)
+          cNonUniform(1:p_OpFM%nNodesForceBlade) = u_OpFM%forceNodesChord(2:p_OpFM%nNodesForceBlade+1)
+
+          ! compute a = c/ds
+          cByS(:) = cNonUniform(:)/sNonUniform(:)
+
+          ! check how a = c/s varies along the span and take its rms to check
+          ! convergence.
+          e(:) = cByS(2:p_OpFM%nNodesForceBlade) - cByS(1:p_OpFM%nNodesForceBlade-1)
+          eSum = sqrt(sum(e(:) * e(:)))
+
+          ! increment the iteration counter
+          counter = counter + 1
+
+       end do
+
+       print*, "Using chord-scaled blade force node clustering."
```

Comment:
@andrew-platt is this appropriate or required to use `ErrMsg` and `ErrStat`?

Reply:
Can you point me to an example of how to do that.  I just want some indication to the user of what method it used, and if it used chord-clustered, some statistics about the final result.

Reply:
I think changing this to `WrScr("Using chord-scaled blade force node clustering"` is sufficient.

Alternatively, you could set `ErrStat = ErrID_Info` and set `ErrMsg` to the text, preferably with the `SetErrStat` subroutine.

We occasionally run into issues with `print` when coupling to other codes (I have vague memories of weirdness with SimuLink).

Reply:
How about this other print statement

WrScr("   -converged to ", eSum, " in ", counter," iterations.")

That won't compile.  How can I do something like that that mixes strings and numbers?

Reply:
The `WrScr` can only accept a single string.  So try the following:

`WrScr(" -converged to "//trim(Num2LStr(eSum))//" in "//trim(Num2LStr(counter))//" iterations.")`

Reply:
That works!  Thanks, Andy.